File: objctemplate/objctemplate/classes/service/TMPJReportService.m
//
//  TMPJReportService.m
//  objctemplate
//
//  Created by supertext on 15/3/11.
//  Copyright (c) 2015年 icegent. All rights reserved.
//

#import "TMPJReportService.h"
//#import "Flurry.h"
inline void TMPJReportAction(NSString *actonName)
{
    [[TMPJReportService defaultService] reportAction:actonName];
}

inline void TMPJReportActionParams(NSString *actonName,NSDictionary *params)
{
    [[TMPJReportService defaultService] reportAction:actonName params:params];
}
@interface TMPJReportService()
@property(nonatomic,readwrite)TMPJReportPlatformOptions platforms;
@end
@implementation TMPJReportService

#pragma mark lifeCircle

+(instancetype)defaultService
{
    static dispatch_once_t onceToken;
    static TMPJReportService *_defaultService;
    dispatch_once(&onceToken, ^{
        _defaultService=[[super allocWithZone:nil] init];
    });
    return _defaultService;
}
+(id)allocWithZone:(NSZone *)zone
{
    return [self defaultService];
}
- (instancetype)init
{
    self = [super init];
    if (self) {
        self.platforms=TMPJReportPlatformFlurry;
        if (self.platforms&TMPJReportPlatformFlurry) {
            //[Flurry startSession:@"2Y6GB4M2B8H2NNHNYCX4"];
        }
#ifdef DEBUG
            //[Flurry setDebugLogEnabled:YES];
#endif
    }
    return self;
}
-(void)reportModel:(TMPJReportModel *)model
{
    if (model) {
        [self reportAction:model.actionName params:model.actionParams];
    }
}
-(void)reportAction:(NSString *)actionName
{
    [self reportAction:actionName params:nil];
}

-(void)reportAction:(NSString *)actionName params:(NSDictionary *)params
{
    if (self.platforms&TMPJReportPlatformFlurry) {
        //[Flurry logEvent:pointNmae withParameters:params];
    }
    if (self.platforms&TMPJReportPlatformGoogle) {
        //google report code hear
    }
}

-(void)reportError:(NSError *)error errorMessage:(NSString *)message errorName:(NSString *)name
{
    if (self.platforms&TMPJReportPlatformFlurry) {
        //[Flurry logError:name message:message error:error];
    }
    if (self.platforms&TMPJReportPlatformGoogle) {
        //google report code hear
    }
}

-(void)reportException:(NSException *)exception exceptionMsg:(NSString *)message exceptionName:(NSString *)name
{
    if (self.platforms&TMPJReportPlatformFlurry) {
        //[Flurry logError:name message:message exception:exception];
    }
    if (self.platforms&TMPJReportPlatformGoogle) {
        //google report code hear
    }
}
@end

File: swifttemplate/swifttemplate/classes/service/TMPJReportService.swift
//
//  TMPJReportService.swift
//  swifttemplate
//
//  Created by supertext on 15/3/11.
//  Copyright © 2016年 icegent. All rights reserved.
//

import EasyTools

struct TMPJReportPlatformOptions : OptionSet {
    var rawValue:UInt = 0
    init(rawValue:UInt){self.rawValue = rawValue}
    init(nilLiteral:()){rawValue = 0}
    static var None:TMPJReportPlatformOptions = [];
    static var Flury:TMPJReportPlatformOptions{
        return self.init(rawValue:1<<1)
    }
    static var Google:TMPJReportPlatformOptions{
        return self.init(rawValue:1<<2)
    }
    static var All:TMPJReportPlatformOptions = [Flury,Google];
}

func TMPJReportAction(_ actionName:String)
{
    TMPJReportService.sharedService.reportAction(actionName);
}
func TMPJReportActionParams(_ actionName:String,actionParams:[String:String]?)
{
    TMPJReportService.sharedService.reportAction(actionName, actionParams: actionParams);
}
final class TMPJReportService: NSObject {
    fileprivate var platforms:TMPJReportPlatformOptions;
    static let sharedService = TMPJReportService();
    fileprivate override init() {
        self.platforms = .All;
        super.init();
    }
    func reportModel(_ model:TMPJReportModel)
    {
        reportAction(model.actionName, actionParams: model.actionParams);
    }
    func reportAction(_ actionName:String)
    {
        reportAction(actionName, actionParams: nil);
    }
    func reportAction(_ actionName:String,actionParams:[String:String]?)
    {
        
    }
}
